### DISCUSSION TITLE
Row Metadata

### DISCUSSION THREAD

**Discussion by vincode-io** [vincode-io/Zavala#131] [Ideas]
Let's discuss how you use outline metadata or columns. Please read [Row Metadata](https://zavala.vincode.io/2021/10/29/Row-Metadata-Proposal.html) for more information.

**Comment by latvia234** [comment]
Right now I'm using row metadata for row creation date and last edited date. 

The outline that I use right now will add this automatically for every row. It is only visible when pointing the mouse at the bulletpoint.
![Schermafbeelding 2021-11-14 om 16 01 47](https://user-images.githubusercontent.com/86680397/141687095-a6094d71-c646-4cca-b306-f2dc4dc4a3a1.png)

@vincode-io Did you already think about an option to hide specific keys/columns in row metadata?

**Reply by vincode-io** [reply]
I had thought about hiding all of them at once, but not based on key. I can see how that would be very useful. It has been added to the requirements for when I implement this.

I hadn't even thought of auto-generated metadata. I would think that would have to be hidden by default or things would just be too busy looking. I think showing it on hover is a good solution.

Are you finding the created/update timestamps useful on a row level? It seems more like a curiosity to me than a useful feature. I wonder what I am missing.

**Reply by latvia234** [reply]
> I hadn't even thought of auto-generated metadata.

When you think about it now, what kind of metadata would be useful to auto-generate?

> I would think that would have to be hidden by default or things would just be too busy looking. 

Yes, I think that metadata is a really powerful option. But it is important to keep the UI clean. Otherwise it would be hard to read when you add a lot of metadata. An option to hide specific keys would be great.

> I think showing it on hover is a good solution.

Showing it when pointing/hovering the mouse at the bulletpoint?

> Are you finding the created/update timestamps useful on a row level? It seems more like a curiosity to me than a useful feature. I wonder what I am missing.

When working with clients I'm using outlines. Each task represents a line. When working with clients it is useful to know when a task is created/updated.

Idea: In combination with a due and done date you could also calculate the timeline. Maybe this would make it possible to create timeline pdf reports for every client that I have.

**Reply by vincode-io** [reply]
> > I hadn't even thought of auto-generated metadata.
> 
> When you think about it now, what kind of metadata would be useful to auto-generate?

I could see recording timestamps for events that Zavala could know about. Created, updated, and completed are the only ones I could think of right now.

> > I think showing it on hover is a good solution.
> 
> Showing it when pointing/hovering the mouse at the bulletpoint?

Yes, but I probably need a better solution. Hovering doesn't work on iOS devices.
 
> > Are you finding the created/update timestamps useful on a row level? It seems more like a curiosity to me than a useful feature. I wonder what I am missing.
> 
> When working with clients I'm using outlines. Each task represents a line. When working with clients it is useful to know when a task is created/updated.

I get it now. It would be very useful for task tracking.

> Idea: In combination with a due and done date you could also calculate the timeline. Maybe this would make it possible to create timeline pdf reports for every client that I have.

That is the kind of thing I want to enable by capturing structured information at the row level. I don't know if I would build full-on time-tracking capabilities into Zavala. I would rather enable Zavala to do things like time-tracking that I might not have even thought of yet.

In your example, Zavala would be used to capture all that information and it could be extracted using Shortcuts and passed to another application to do specialized reporting. How I envision it working is that you would do all your daily work in Zavala and then when you wanted a report you would fire off a Shortcut. This Shortcut could extract the captured data from Zavala, pass it to Numbers (or whatever), tell Numbers to crunch the data into a report, and then present you the PDF.

It isn't as slick as having everything build into Zavala, but it is much more flexible and customizable.

**Reply by latvia234** [reply]
> > Idea: In combination with a due and done date you could also calculate the timeline. Maybe this would make it possible to create timeline pdf reports for every client that I have.
> 
> That is the kind of thing I want to enable by capturing structured information at the row level. I don't know if I would build full-on time-tracking capabilities into Zavala. I would rather enable Zavala to do things like time-tracking that I might not have even thought of yet.
> 
> In your example, Zavala would be used to capture all that information and it could be extracted using Shortcuts and passed to another application to do specialized reporting. How I envision it working is that you would do all your daily work in Zavala and then when you wanted a report you would fire off a Shortcut. This Shortcut could extract the captured data from Zavala, pass it to Numbers (or whatever), tell Numbers to crunch the data into a report, and then present you the PDF.

That's how I was thinking about it also, but forgot to mention that I would like to use Shortcuts for this idea.

> It isn't as slick as having everything build into Zavala, but it is much more flexible and customizable.

That's no problem. I totally agree with this. :-) Please keep Zavala as simple as possible.

**Comment by latvia234** [comment]
@vincode-io Do you think that conditional formatting would be useful in combination with row metadata?

When I can create a hidden key called "important" I'll fill it with a certain phrase or number. And the row will be formatted in a certain way.

Right now I use it with my collegue, when he's commenting on me, he'll add a certain color. Of course we can also use tags, but sometimes I just like to use colors instead of tags.

![Schermafbeelding 2021-11-15 om 21 15 46](https://user-images.githubusercontent.com/86680397/141848347-b9456329-d00e-432a-95e7-436e774feb3f.png)

**Reply by latvia234** [reply]
**Conditional emoji's & id's**
I wrote some macro's for Keyboard Maestro to link to other software via id's. (Maybe this can also be done with Shortcuts?)

In the outline I write the subject title for an email.

When typing the subject I also type #Email and automatically the id tag appears after #Email: #48e8ba 

So right now this is what you can see in the outline row: Subject with Zavala row id #Email #48e8ba

Typing #Email will trigger a KM macro, and the email application will appear with prefilled subject. Inside the subject you can find the Zavala row id:
![Schermafbeelding 2021-11-15 om 22 05 20](https://user-images.githubusercontent.com/86680397/141854233-622e235e-346e-4c0f-9aa6-606f91425b79.png)

Advantages:
- When I'm in the outline on another day, I start the macro by pressing some keys and it will start the email application with the linked subject. So I don't have to search manually for the email anymore.
- When I'm in the email app on another day, I start the macro and it will open the outline at the row with id. I don't have to search inside the outline, I can start it when I'm in the e-mail app.
- This can work for a lot of other apps

Questions:
- How could this be done with Zavala? Do you think that I can use a hidden keys named "id" and "Linked" for this?
- Can you make it possible to enable emoji's to show if a hidden key contains specific information?
   - If hidden key "Linked" contains text "Email" show "📨"
   - If hidden key "Linked" contains text "WhatsApp" show "💬"
   - If hidden key "Linked" contains text "Blog" show "📝"


![Schermafbeelding 2021-11-15 om 22 11 34](https://user-images.githubusercontent.com/86680397/141854424-72130e32-9f28-4d1f-89c7-7a7290eec5f9.png)


It would be great if clicking on the emoji and/or a keyboard shortcut could bring you to the linked application.

**Reply by vincode-io** [reply]
Is this using URL schemes to link applications?

**Reply by latvia234** [reply]
No, just a Keyboard Maestro macro using keyboard shortcuts. But I think that using URL schemes could also work, if the other application does support URL schemes.

**Reply by vincode-io** [reply]
I'm not familiar with Keyboard Maestro, so I don't really understand how you are linking applications together with it. I would think that it has to be using URL schemes.

Zavala is designed to work with other applications by linking using URL schemes. It doesn't do it at the row level. Just at the document level. It would be possible to implement deep linking to the row level though, I just haven't gotten any requests for it.

To deep link to an outline in Zavala, you can get the document link in the Documents View (the middle column) using the context menu. Just paste that into an application (Notes works for example) that supports links and you have a clickable link to a Zavala outline.

Zavala also will link other applications that support custom URL schemes or even just regular links to the web. You do it in the Link Dialog. Version 2.0 of Zavala adds full support for emoji in the outlines, so when that comes out you could create an emoji that links to what ever you would like.

**Comment by vincode-io** [comment]
I'm not sure about conditional formatting. I definitely see how it would be useful, I'm just concerned about the complexity of configuring it to work. It might be a bit farther than I want to go. I'll keep thinking about it and see if I can come up with something not complicated.

**Reply by latvia234** [reply]
It's very important to keep Zavala as simple as possible. So I understand your decision. It's better to think about it and not add features that will clutter the UI.

I remember the mission statement and I think it's good to read that before implement a new feature.

**Comment by tmanternach** [comment]
I just came across your post over on scripting/drummerCMS. This app is great! Am I missing an obvious step here though? Does Zavala 1.2.1 (macOS) support *creating* Row Metadata? I read your post about it, but it might just for read-only for now? Just wanted to see if I was missing something obvious.

**Reply by vincode-io** [reply]
Right now it is just a proposal for a feature. Sorry, you aren't missing anything. It just hasn't been implemented yet.

**Reply by tmanternach** [reply]
No problem. Keep up the great work here!

**Comment by damienzhang** [comment]
I'm thinking row meta data that refers to other rows could be super useful in, e.g., project planning. Say you have a list of tasks that depend on each other:

row1. buy ingredients
row2. bake cake [requires: row1]
row3. invite guests

Then you can imagine with a search of (rows with no [requires] OR rows where [required] is completed), you have a list of the tasks that you can actually start on. Another search of (rows where [required] is not completed) shows the list of blocked tasks.

**Comment by MahatmaFatalError** [comment]
what is the current status of the this feature?

**Reply by vincode-io** [reply]
Right now this feature is on hold. My current focus is to make Zavala more stable and usable without adding much more to the UI. This is a big change and wouldn't happen until the next major release at the earliest.
